### ISSUE TITLE
nexpy does not read nexus files created by pni-libraries

### ISSUE THREAD

**Issue by jkotan** [nexpy/nexpy#146]
Hi,

I've just tried to use nexpy to read nexus files created by PNI libraries https://pni-libraries.github.io/python-pni/nexus_files.html. 
Unfortunately, nexpy cannot read them, i.e. it has problems with reading string attributes written as an one element array (it is a standard for libpniio). Nexpy shows an error about problems with the NXroot group during opening the nexus file. 
An example of the pni nexus file can be found at https://github.com/jkotan/file-examples/tree/master/pninexus.

Is it possible to fix it?

Best regards,
Jan

**Comment by rayosborn** [comment]
Thank you for the report. Ironically, the problem is that the NeXus class of the top level group is explicitly defined by the attribute, i.e, NX_class=('NXroot',), but the size-1 array prevented it from being recognized as the NXroot class. Most files don't set the NXroot class at all. The API can handle size-1 arrays, but the group class was determined before the conversion to a scalar string. It's easy to fix, although there may be other implications that I should check. I was going to issue a pre-release version soon and will let you know when there's something to test.

**Comment by jkotan** [comment]
Thanks.

**Comment by rayosborn** [comment]
I have merged a fix for this issue (nexpy/nexusformat#61) into the [nexusformat v0.4.18 branch](/nexpy/nexusformat/tree/v0.4.18). Note that the bug is in [nexusformat package](/nexpy/nexusformat), which NeXpy uses to read and write NeXus files, not NeXpy itself. If you are able to test this branch, please let me know if it works.  

I was able to open your example file and produce the following tree:
```
myys_00003:NXroot
  @HDF5_version = '1.10.0'
  @NeXus_version = '4.3.0'
  @file_name = '/tmp/myys_00003.nxs'
  @file_time = '2018-04-05T11:55:15.014928+0200'
  @file_update_time = '2018-04-05T11:55:20.957440+0200'
  nexus_logs:NXcollection
    configuration:NXcollection
      nexus__entry__1_xml = '<?xml version="1.0" ?><definition> ...'
      python_version = '2.7.13 (default, Nov 24 2017, 17:33:09) ...'
  scan:NXentry
    data:NXdata
      exp_c01 = float64(11)
        @nexdatas_source = '<datasource type="TANGO" name="exp_c01">      ...'
        @nexdatas_strategy = 'STEP'
        @type = 'NX_FLOAT64'
      exp_c02 = float64(11)
        @nexdatas_source = '<datasource type="TANGO" name="exp_c02">      ...'
        @nexdatas_strategy = 'STEP'
        @type = 'NX_FLOAT64'
      exp_mot03 = float64(11)
        @nexdatas_source = '<datasource type="CLIENT" name="exp_mot03">   ...'
        @nexdatas_strategy = 'STEP'
        @type = 'NX_FLOAT64'
      point_nb = int64(11)
        @nexdatas_source = '<datasource type="CLIENT" name="point_nb">    ...'
        @nexdatas_strategy = 'STEP'
        @type = 'NX_INT64'
      timestamp = float64(11)
        @nexdatas_source = '<datasource type="CLIENT" name="timestamp">   ...'
        @nexdatas_strategy = 'STEP'
        @type = 'NX_FLOAT64'
    end_time = '2018-04-05T11:55:20.929439+0200'
      @nexdatas_source = '<datasource type="CLIENT" name="end_time">    ...'
      @nexdatas_strategy = 'FINAL'
      @type = 'NX_DATE_TIME'
    experiment_identifier = 'P64_2018-04-05T11:55:14.632973+0200@haso228jk'
      @nexdatas_source = '<datasource type="PYEVAL" name="beamtime_id"> ...'
      @nexdatas_strategy = 'INIT'
      @type = 'NX_CHAR'
    instrument:NXinstrument
      collection:NXcollection
        exp_mot03 = float64(11)
          @nexdatas_source = '<datasource type="CLIENT" name="exp_mot03">   ...'
          @nexdatas_strategy = 'STEP'
          @type = 'NX_FLOAT64'
        point_nb = int64(11)
          @nexdatas_source = '<datasource type="CLIENT" name="point_nb">    ...'
          @nexdatas_strategy = 'STEP'
          @type = 'NX_INT64'
        timestamp = float64(11)
          @nexdatas_source = '<datasource type="CLIENT" name="timestamp">   ...'
          @nexdatas_strategy = 'STEP'
          @type = 'NX_FLOAT64'
      detector:NXdetector
        collection:NXcollection
          exp_c01 = float64(11)
            @nexdatas_source = '<datasource type="TANGO" name="exp_c01">      ...'
            @nexdatas_strategy = 'STEP'
            @type = 'NX_FLOAT64'
          exp_c02 = float64(11)
            @nexdatas_source = '<datasource type="TANGO" name="exp_c02">      ...'
            @nexdatas_strategy = 'STEP'
            @type = 'NX_FLOAT64'
      name = 'P64 Advanced X-Ray Absorption Spectroscopy'
        @nexdatas_strategy = 'INIT'
        @short_name = 'P64'
        @type = 'NX_CHAR'
      source:NXsource
        name = 'PETRA III'
          @nexdatas_strategy = 'INIT'
          @short_name = 'PETRAIII'
          @type = 'NX_CHAR'
        probe = 'x-ray'
          @nexdatas_strategy = 'INIT'
          @type = 'NX_CHAR'
        type = 'Synchrotron X-ray Source'
          @nexdatas_strategy = 'INIT'
          @type = 'NX_CHAR'
    program_name = 'NexDaTaS'
      @configuration = '2.8.2.P64@HASNP64.6659'
      @scan_command = 'ascan exp_mot03 0.0 10.0 10 0.01'
      @scan_id = '3'
      @type = 'NX_CHAR'
      @version = '2.18.0'
    sample:NXsample
      chemical_formula = ''
        @nexdatas_source = '<datasource type="CLIENT" name="chemical_formu...'
        @nexdatas_strategy = 'INIT'
        @type = 'NX_CHAR'
      name = ''
        @nexdatas_source = '<datasource type="CLIENT" name="sample_name"> ...'
        @nexdatas_strategy = 'INIT'
        @type = 'NX_CHAR'
    start_time = '2018-04-05T11:55:14.632973+0200'
      @nexdatas_source = '<datasource type="CLIENT" name="start_time">  ...'
      @nexdatas_strategy = 'INIT'
      @type = 'NX_DATE_TIME'
    title = ''
      @nexdatas_source = '<datasource type="CLIENT" name="title">    ...
      @nexdatas_strategy = 'INIT'
      @type = 'NX_CHAR'
```
As @prjemian guessed, the attributes did cause a few problems since they are so long, but I trimmed the output to accommodate them in the same merge. There are no signal or axes attributes in the NXdata group, and the `type` attributes are, I believe, completely redundant - at least, they are not used by the Python API, but the file is readable.

**Comment by jkotan** [comment]
Great, Thanks. I've done short tests and it works. Also for a file containing `@signal` the corresponding plot is displayed. I will do more checks tomorrow.

**Comment by vasole** [comment]
@rayosborn 

In my opinion we should not support that. Those attributes are defined as strings (variable length or not) and not as arrays of strings.

**Comment by vasole** [comment]
Here you can see how **all** attributes are arrays of strings instead of strings in those files (square brackets everywhere)

![badattrs](https://user-images.githubusercontent.com/5636251/38648606-6aeb0d00-3df2-11e8-9d82-c0cf05c33877.png)

**Comment by rayosborn** [comment]
That might be something for the NIAC to discuss. My goal with NeXpy is to read NeXus files that are out there, and to try to accommodate the various ways they may have been written. In any case, I believe that the NeXus C-API writes, or used to write, strings as size-1 string arrays, producing the output that you illustrated. Unless the NIAC decides to explicitly forbid such arrays, I have to read them. It's not that hard to do, but I would think it's a headache for anyone writing their own readers in h5py.

**Comment by vasole** [comment]
The NeXus C-API was writing strings as strings and not as arrays because files produced with it were producing the output without the square brackets. I had to encounter a file written with pninexus to find an example of **official** file not behaving that way.

I would really like to know how many NIAC members are producing generic software to deal with NeXus files from different facilities because I guess they would have thought twice about allowing that relaxation of the rule.

Hey, I even refuse to read SPEC files in PyMca if they have modified the line endings from LF to CR/LF. Yes it would cost nothing to support them but they would not any longer be SPEC files.

**Comment by rayosborn** [comment]
Coincidentally, I raised this very issue at the last telco, and I will suggest it is on the agenda for the upcoming NIAC. As you suggest, it is an issue that those of us writing generic software are keenly aware of. However, we will always have a legacy of files going back many years - backward-compatibility is always tough.

**Comment by rayosborn** [comment]
Fixed by nexpy/nexusformat#61.
